containerd-shim-v2/service.go

```diff
@@ -202,9 +215,35 @@ func (s *service) StartShim(ctx context.Context, id, containerdBinary, container
 	return address, nil
 }
 
-func (s *service) forward(publisher events.Publisher) {
+func (s *service) forward(l events.Publisher) {
+	var addr string
+	var binaryPath string
+
+	// Since containerd shim doesn't export
+	//    type remoteEventsPublisher struct {
+	//         address              string
+	//         containerdBinaryPath string
+	//    }
+	// thus it's needed to get the struct's
+	// fields value from an interface value
+	// using reflection.
+	elem := reflect.ValueOf(l).Elem()
+	typeOfT := elem.Type()
+
+	for i := 0; i < elem.NumField(); i++ {
+		f := elem.Field(i)
+		switch typeOfT.Field(i).Name {
+		case "address":
```

Comment:
This might be a bit "too clever" ;) Since those struct elements are private, they could change and should be able to change without breaking any clients. But we would be broken here if they changed if I'm understanding this correctly?

Reply:
It might be worth raising a PR on containerd shim to see if they would be prepared to make these fields public.

Reply:
Yes, to fix this issue gently there must be some adjustments in containerd side, by now I think we can workaround it in our side first and then we communicate with containerd community on how to fix it ultimately.

Reply:
I'd really appreciate if we could start the conversation with containerd community right away to get a hint at how they feel regarding this proposal. I don't want to merge to Kata Containers a workaround that might never be properly fixed if containerd community does not agree on these changes.
Please start the discussion with them by adding @jodh-intel and myself in copy!

Reply:
Agree. Please open an issue with containerd for this @lifupan.